metabase/metabase
# Join performance issues

Changes:
- Optimize MLv2 calls. Defer as much as possible. 
- Remove effects and any async logic. 
- Split `JoinStep` into `Join` and `JoinDraft`. Draft version is used for incomplete joins (without the table / left-hand-side / right-hand-side column). Draft joins have different state, and all updates are done differently. **It should be much cleaner and easier to follow than before**. If in doubt, compare with `use-join`, `use-join-condition` and `JoinStep` from `master`. Because of the simplification I could do MLv2 calls in an optional way for each case - complete or draft join/condition.

I'd recommend reviewing from `JoinStep`. In files that are changed heavily the diff is irrelevant. 

How to verify:
- Add a question with a join
- Open the column picker for the joined table
- Start React profiler
- Tick or untick or a column
- `Notebook` should re-render only once

Currently there are 2 rerenders, however the notebook is re-rendered indeed once.
1st:
<img width="920" alt="Screenshot 2024-03-06 at 09 09 25" src="https://github.com/metabase/metabase/assets/8542534/8c90b8f1-836a-445b-8f39-9ba9b6b0fbec">

2nd:
<img width="1728" alt="Screenshot 2024-03-06 at 09 08 13" src="https://github.com/metabase/metabase/assets/8542534/db768b60-1d70-4a48-a502-589e4c00b6e7">
